web/src/user_profile.js

```diff
@@ -422,6 +435,183 @@ function handle_remove_stream_subscription(target_user_id, sub, success, failure
+        is_incoming_webhook_bot: bot.bot_type === INCOMING_WEBHOOK_BOT_TYPE,
+    });
+    $container.append(html_body);
+    let avatar_widget;
+
+    const bot_type = bot.bot_type.toString();
+    const service = bot_data.get_services(bot.user_id)[0];
+    edit_bot_post_render();
+    $("#user-profile-modal").on("click", ".dialog_submit_button", () => {
+        const role = Number.parseInt($("#bot-role-select").val().trim(), 10);
+        const $full_name = $("#bot-edit-form").find("input[name='full_name']");
+        const url = "/json/bots/" + encodeURIComponent(bot.user_id);
+
+        const formData = new FormData();
+        formData.append("csrfmiddlewaretoken", csrf_token);
+        formData.append("full_name", $full_name.val());
+        formData.append("role", JSON.stringify(role));
+        const new_bot_owner_id = bot_owner_dropdown_widget.value();
+        if (new_bot_owner_id) {
+            formData.append("bot_owner_id", new_bot_owner_id);
+        }
+
+        if (bot_type === OUTGOING_WEBHOOK_BOT_TYPE) {
+            const service_payload_url = $("#edit_service_base_url").val();
+            const service_interface = $("#edit_service_interface").val();
+            formData.append("service_payload_url", JSON.stringify(service_payload_url));
+            formData.append("service_interface", service_interface);
+        } else if (bot_type === EMBEDDED_BOT_TYPE && service !== undefined) {
+            const config_data = {};
+            $("#config_edit_inputbox input").each(function () {
+                config_data[$(this).attr("name")] = $(this).val();
+            });
+            formData.append("config_data", JSON.stringify(config_data));
+        }
+
+        const $file_input = $("#bot-edit-form").find(".edit_bot_avatar_file_input");
+        for (const [i, file] of Array.prototype.entries.call($file_input[0].files)) {
+            formData.append("file-" + i, file);
+        }
+
+        const $submit_btn = $("#user-profile-modal .dialog_submit_button");
+        const $cancel_btn = $("#user-profile-modal .dialog_exit_button");
+        settings_users.show_button_spinner($submit_btn);
+        $cancel_btn.prop("disabled", true);
+
+        channel.patch({
+            url,
+            data: formData,
+            processData: false,
+            contentType: false,
+            success() {
+                avatar_widget.clear();
+                hide_user_profile();
+            },
+            error(xhr) {
+                ui_report.error(
+                    $t_html({defaultMessage: "Failed"}),
+                    xhr,
+                    $("#bot-edit-form-error"),
+                );
+                // Scrolling modal to top, to make error visible to user.
+                $("#bot-edit-form")
+                    .closest(".simplebar-content-wrapper")
+                    .animate({scrollTop: 0}, "fast");
+                settings_users.hide_button_spinner($submit_btn);
```

Comment:
We already have an issue to move this to `loading.ts` probably, so this can be handled separately.

Reply:
Yeah that seems fine as a follow-up.